components/eamxx/src/physics/iop_forcing/eamxx_iop_forcing_process_interface.cpp

```diff
@@ -0,0 +1,546 @@
+  const auto policy_iop = ESU::get_default_team_policy(m_num_cols, nlev_packs);
+
+  // Reset internal WSM variables.
+  m_workspace_mgr.reset_internals();
+
+  // Avoid implicit capture of this
+  auto wsm = m_workspace_mgr;
+  auto num_levs = m_num_levs;
+
+  // Apply IOP forcing
+  Kokkos::parallel_for("apply_iop_forcing", policy_iop, KOKKOS_LAMBDA (const MemberType& team) {
+    const int icol  =  team.league_rank();
+
+    auto ps_i = ps(icol);
+    auto u_i = Kokkos::subview(horiz_winds, icol, 0, Kokkos::ALL());
+    auto v_i = Kokkos::subview(horiz_winds, icol, 1, Kokkos::ALL());
+    auto T_mid_i = ekat::subview(T_mid, icol);
+    auto qv_i = ekat::subview(qv, icol);
+    auto Q_i = Kokkos::subview(Q, icol, Kokkos::ALL(), Kokkos::ALL());
+
+    auto ws = wsm.get_workspace(team);
+    uview_1d<Pack> ref_p_mid, ref_p_int, ref_p_del;
+    ws.take_many_contiguous_unsafe<3>({"ref_p_mid", "ref_p_int", "ref_p_del"},
+                                      {&ref_p_mid,  &ref_p_int,  &ref_p_del});
+
+    // Compute reference pressures and layer thickness.
+    // TODO: Allow geometry data to allocate packsize
+    auto s_ref_p_mid = ekat::scalarize(ref_p_mid);
+    auto s_ref_p_int = ekat::scalarize(ref_p_int);
+    Kokkos::parallel_for(Kokkos::TeamVectorRange(team, num_levs+1), [&](const int& k) {
+      s_ref_p_int(k) = hyai(k)*ps0 + hybi(k)*ps_i;
+      if (k < num_levs) {
+        s_ref_p_mid(k) = hyam(k)*ps0 + hybm(k)*ps_i;
+      }
+    });
+    team.team_barrier();
+    ColOps::compute_midpoint_delta(team, num_levs, ref_p_int, ref_p_del);
+    team.team_barrier();
+
+    if (iop_dosubsidence) {
+    // Compute subsidence due to large-scale forcing
+      advance_iop_subsidence(team, num_levs, dt, ps_i, ref_p_mid, ref_p_int, ref_p_del, omega, ws, u_i, v_i, T_mid_i, Q_i);
+    }
+
+    // Update T and qv according to large scale forcing as specified in IOP file.
+    advance_iop_forcing(team, num_levs, dt, divT, divq, T_mid_i, qv_i);
+
+    if (iop_coriolis) {
+      // Apply coriolis forcing to u and v winds
+      iop_apply_coriolis(team, num_levs, dt, target_lat, u_ls, v_ls, u_i, v_i);
+    }
+
+    // Release WS views
+    ws.release_many_contiguous<3>({&ref_p_mid, &ref_p_int, &ref_p_del});
+  });
+  Kokkos::fence();
+
+  // Nudge the domain based on the domain mean
+  // and observed quantities of T, Q, u, and v
+  if (iop_nudge_tq or iop_nudge_uv) {
+    // Compute domain mean of qv, T_mid, u, and v
+    const auto qv_mean = m_buffer.qv_mean;
+    const auto t_mean  = m_buffer.t_mean;
+    const auto u_mean  = m_buffer.u_mean;
+    const auto v_mean  = m_buffer.v_mean;
+
+    const auto qv_mean_h = Kokkos::create_mirror_view(qv_mean);
+    const auto t_mean_h  = Kokkos::create_mirror_view(t_mean);
+    const auto u_mean_h  = Kokkos::create_mirror_view(u_mean);
+    const auto v_mean_h  = Kokkos::create_mirror_view(v_mean);
+
+    const auto num_global_cols = m_grid->get_num_global_dofs();
+    for (int k=0; k<m_num_levs; ++k) {
```

Comment:
Have you thought about doing an outer || for? Yes, the access pattern on GPU is screwed up, but that's the same here. You may as well throw in another layer of ||ism, and do all calculations on device. It will at least save you from creating the mirrot view and do deep copies after the vertical for loop...

Reply:
The reason I did not is that we do an MPI_AllReduce within the loop and I assumed that couldn't be done on device. Is that a correct assumption?

Anyways, I could save that computation for outside the nest ||for/reduce and all reduce on the entire view, but I'll still need the mirrors I think?

Reply:
You could do the rescaling at the end of the nested for/reduce, no? Since the scaling is num_global_cols for all entries, no?

Reply:
Yes, but wouldn't I still need the host mirrors to do the all reduce? And I guess the scaling as well?

Reply:
Yeah, you would, unless SCREAM_MPI_ON_DEVICE is defined. But anyways, I think the convo regarding the Field utility below may be overriding this one..